src/python/platforms/fuchsia/device.py

```diff
@@ -54,98 +53,140 @@ def setup_qemu_values(initial_setup=True):
+    """
+    qemu_vars = _fetch_qemu_vars()
+
+    # Get a free port for the VM, so we can SSH in later.
+    tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    tcp.bind(('localhost', 0))
+    _, port = tcp.getsockname()
+    tcp.close()
+    # Fuzzing jobs that SSH into the QEMU VM need access to this env var.
+    environment.set_value('FUCHSIA_PORTNUM', port)
+    environment.set_value('FUCHSIA_RESOURCES_DIR',
+                          qemu_vars['fuchsia_resources_dir'])
+
+    # yapf: disable
+    qemu_args = [
+        '-m', '2048',
+        '-nographic',
+        '-kernel', qemu_vars['kernel_path'],
+        '-initrd', qemu_vars['initrd_path'],
+        '-smp', '4',
+        '-drive',
+        ('file=' + qemu_vars['drive_path'] + ',format=raw,if=none,'
+         'id=blobstore'),
+        '-device', 'virtio-blk-pci,drive=blobstore',
+        '-monitor', 'none',
+        '-append', '"kernel.serial=legacy TERM=dumb"',
+        '-machine', 'q35',
+        '-display', 'none',
+        '-netdev',
+        ('user,id=net0,net=192.168.3.0/24,dhcpstart=192.168.3.9,'
+         'host=192.168.3.2,hostfwd=tcp::') + str(port) + '-:22',
+        '-device', 'e1000,netdev=net0,mac=52:54:00:63:5e:7b',
+        '-L', qemu_vars['sharefiles_path']
+    ]
+    # yapf: enable
+
+    # Detecing KVM is tricky, so use an environment variable to determine
+    # whether to turn it on or not.
+    if environment.get_value('FUCHSIA_USE_KVM'):
+      qemu_args.extend(['-cpu', 'host,migratable=no'])
+      qemu_args.append('-enable-kvm')
+    else:
+      # Can't use host CPU since we don't necessarily have KVM on the machine.
+      # Emulate a Haswell CPU with a few feature toggles. This mirrors the most
+      # common configuration for Fuchsia VMs when using in-tree tools.
+      qemu_args.extend(['-cpu', 'Haswell,+smap,-check,-fsgsbase'])
+
+    # Get the list of fuzzers for ClusterFuzz to choose from.
+    host = Host.from_dir(
+        os.path.join(qemu_vars['fuchsia_resources_dir'], 'build', 'out',
+                     'default'))
+    Device(host, 'localhost', str(port))
+    Fuzzer.filter(host.fuzzers, '')
+
+    # Fuzzing jobs that SSH into the QEMU VM need access to this env var.
+    environment.set_value('FUCHSIA_PKEY_PATH', qemu_vars['pkey_path'])
+    logs.log('Ready to run QEMU. Command: ' + qemu_vars['qemu_path'] + ' ' +
+             str(qemu_args))
+    self.process_runner = new_process.ProcessRunner(qemu_vars['qemu_path'],
+                                                    qemu_args)
+
+  def run(self):
+    """Actually runs a QEMU VM, assuming `create` has already been called."""
+    if not self.process_runner:
+      raise QemuException('Attempted to `run` QEMU VM before calling `create`')
+    self.popen = self.process_runner.run(
+        stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    time.sleep(_QEMU_WAIT_SECONDS)
+
+  def kill(self):
+    if not self.popen:
```

Comment:
nit: docstring. also maybe raise an exception instead?

Reply:
added a docstring.

presumably if someone's calling "kill", they just want the process good and dead, and we shouldn't throw an error if it happens to already be dead?  (i notice terminate_processes_matching_names just ignores psutil.NoSuchProcess errors, for example)

Reply:
sure, that seems reasonable.